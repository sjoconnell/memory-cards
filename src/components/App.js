import React, { Component } from "react";
import Card from "./Card";
import createCards from "../util/cardUtil";
import "./styles/App.scss";

class App extends Component {
  state = {
    flippedCardIds: [],
    matchedCardIds: [],
    freezeGame: false,
    cards: createCards()
  };

  areCardsAMatch = id => {
    const firstCardFlipped = this.state.cards.find(
      card => card.id === this.state.flippedCardIds[0]
    );
    const secondCardFlipped = this.state.cards.find(card => card.id === id);
    return secondCardFlipped.value === firstCardFlipped.value;
  };

  resetUnmatchedCards = () => {
    this.setState({
      flippedCardIds: [],
      freezeGame: false
    });
  };

  handleCardClick = id => {
    this.setState({
      freezeGame: true
    });
    // If no cards are flipped, flip the card
    if (this.state.flippedCardIds.length === 0) {
      this.setState({
        flippedCardIds: [...this.state.flippedCardIds, id],
        freezeGame: false
      });
    } else {
      // check if the clicked card is the same as the currently flipped card
      // This should not happen as we disable the click of flipped card but here just in case
      if (this.state.flippedCardIds.includes(id)) {
        this.setState({
          freezeGame: false
        });
      } else {
        // flip the card as it is not the same as the currently flipped card
        this.setState({
          flippedCardIds: [...this.state.flippedCardIds, id]
        });
        // Check if both flipped cards are match
        if (this.areCardsAMatch(id)) {
          this.setState({
            freezeGame: false,
            matchedCardIds: [
              ...this.state.matchedCardIds,
              id,
              this.state.flippedCardIds[0]
            ],
            flippedCardIds: []
          });
        } else {
          // reset both flipped cards since they are not a match
          setTimeout(this.resetUnmatchedCards, 1750);
        }
      }
    }
  };

  render() {
    return (
      <div className="game">
        {this.state.cards.map(card => (
          <Card
            value={card.value}
            id={card.id}
            key={card.id}
            flipped={this.state.flippedCardIds.includes(card.id)}
            matched={this.state.matchedCardIds.includes(card.id)}
            disabled={this.state.freezeGame}
            handleCardClick={this.handleCardClick}
          />
        ))}
      </div>
    );
  }
}

export default App;
