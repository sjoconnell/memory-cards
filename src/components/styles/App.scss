.game {
  max-width: 670px;
  min-height: 400px;
  display: flex;
  flex-wrap: wrap;
  background-color: blue;
  justify-content: space-around;
  padding: 20px;
  perspective: 1000px;
}
