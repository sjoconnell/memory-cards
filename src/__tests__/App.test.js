import React from "react";
import ReactDOM from "react-dom";
import { create } from "react-test-renderer";
import App from "../components/App";

jest.mock("../util/cardUtil.js", () => () => {
  const cardValues = [
    "😃",
    "💩",
    "🤡",
    "🐵",
    "🏀",
    "🌴",
    "👍",
    "👻",
    "🔥",
    "🍕",
    "🏠",
    "🐢"
  ];
  let id = 0;

  return cardValues.reduce(
    (accumulator, value) => [
      ...accumulator,
      { id: id++, value },
      { id: id++, value }
    ],
    []
  );
});

test("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test("App snapshot", () => {
  const c = create(<App />);
  expect(c.toJSON()).toMatchSnapshot();
});

test("first card flipped adds id to flippedCardsIds", () => {
  const c = create(<App />);
  const instance = c.getInstance();

  expect(instance.state.flippedCardIds.length).toBe(0);
  instance.handleCardClick(5);
  expect(instance.state.flippedCardIds.length).toBe(1);
  expect(instance.state.flippedCardIds.includes(5)).toBe(true);
});

test("clicking on a flipped card does not add to flippedCardsIds", () => {
  const c = create(<App />);
  const instance = c.getInstance();

  expect(instance.state.flippedCardIds.length).toBe(0);
  instance.handleCardClick(7);
  expect(instance.state.flippedCardIds.length).toBe(1);
  instance.handleCardClick(7);
  expect(instance.state.flippedCardIds.length).toBe(1);
});

test("matched cards' ids are added to matchedCardIds", () => {
  const c = create(<App />);
  const instance = c.getInstance();

  expect(instance.state.matchedCardIds.length).toBe(0);

  const matchingCards = instance.state.cards
    .filter(card => card.value === "🔥")
    .map(matchedCards => matchedCards.id);

  matchingCards.forEach(card => instance.handleCardClick(card));

  expect(instance.state.matchedCardIds.length).toBe(2);
  expect(
    matchingCards.every(card => instance.state.matchedCardIds.includes(card))
  ).toBe(true);
  expect(instance.state.flippedCardIds.length).toBe(0);
});

test("flipped cards that don't match reset flippedCardIds", async () => {
  const c = create(<App />);
  const instance = c.getInstance();

  const unmatchedCard1 = instance.state.cards.find(card => card.value === "🔥");
  const unmatchedCard2 = instance.state.cards.find(card => card.value === "🌴");

  expect(instance.state.flippedCardIds.length).toBe(0);
  instance.handleCardClick(unmatchedCard1.id);
  expect(instance.state.flippedCardIds.length).toBe(1);
  instance.handleCardClick(unmatchedCard2.id);
  await new Promise(resolve => setTimeout(resolve, 2000));
  expect(instance.state.flippedCardIds.length).toBe(0);
  expect(instance.state.matchedCardIds.length).toBe(0);
});
