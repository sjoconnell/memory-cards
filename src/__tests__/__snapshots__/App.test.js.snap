// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<div
  className="game"
>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        😃
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        😃
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        💩
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        💩
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🤡
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🤡
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🐵
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🐵
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🏀
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🏀
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🌴
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🌴
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        👍
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        👍
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        👻
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        👻
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🔥
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🔥
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🍕
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🍕
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🏠
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🏠
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🐢
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
  <div
    className="card  "
    onClick={[Function]}
  >
    <div
      className="card-front"
    >
      <span
        aria-label="front-card-icon"
        role="img"
      >
        🐢
      </span>
    </div>
    <div
      className="card-back"
    >
      <span
        aria-label="back-card-icon"
        role="img"
      >
        ❓
      </span>
    </div>
  </div>
</div>
`;
